http_error.go

```diff
@@ -26,45 +27,94 @@ const ErrorHeader = "X-Forwarder-Error"
 var ErrProxyLocalhost = denyError{errors.New("localhost proxying is disabled")}
 
 func errorResponse(req *http.Request, err error) *http.Response {
-	// Get the root cause of the error.
-	rootErr := err
-	for {
-		e := errors.Unwrap(rootErr)
-		if e == nil {
-			break
-		}
-		rootErr = e
+	handlers := []errorHandler{
+		handleNetError,
+		handleTLSRecordHeader,
+		handleTLSCertificateError,
+		handleDenyError,
 	}
 
 	var (
-		msg  string
 		code int
+		msg  string
 	)
-	if nerr, ok := err.(net.Error); ok { //nolint // switch is not an option for type asserts
+	for _, h := range handlers {
+		code, msg = h(req, err)
+		if code != 0 {
+			break
+		}
+	}
+	if code == 0 {
+		code = http.StatusInternalServerError
+		msg = "An unexpected error occurred"
+	}
+
+	resp := proxyutil.NewResponse(code, bytes.NewBufferString(msg+"\n"), req)
+	resp.Header.Set(ErrorHeader, err.Error())
+	resp.Header.Set("Content-Type", "text/plain; charset=utf-8")
+	resp.ContentLength = int64(len(msg) + 1)
+	return resp
+}
+
+type errorHandler func(*http.Request, error) (int, string)
+
+func handleNetError(_ *http.Request, err error) (code int, msg string) {
+	if nerr, ok := cause(err).(net.Error); ok {
 		if nerr.Timeout() {
 			code = http.StatusGatewayTimeout
 			msg = "Timed out connecting to remote host"
 		} else {
 			code = http.StatusBadGateway
 			msg = "Failed to connect to remote host"
 		}
-	} else if _, ok := err.(denyError); ok { //nolint:errorlint // makes no sense here
+	}
+
+	return
+}
+
+func handleTLSRecordHeader(_ *http.Request, err error) (code int, msg string) {
+	var headerErr *tls.RecordHeaderError
+	if ok := errors.As(err, &headerErr); ok {
+		code = http.StatusBadGateway
+		msg = "TLS handshake failed"
+	}
+
+	return
+}
+
+func handleTLSCertificateError(_ *http.Request, err error) (code int, msg string) {
+	var certErr *tls.CertificateVerificationError
+	if ok := errors.As(err, &certErr); ok {
+		code = http.StatusBadGateway
+		msg = "TLS handshake failed"
+	}
+
+	return
+}
+
+func handleDenyError(req *http.Request, err error) (code int, msg string) {
+	if _, ok := cause(err).(denyError); ok { //nolint:errorlint // makes no sense here
```

Comment:
we could use `errors.As` here as well

Reply:
It doesn't matter, since we need to keep `cause(err)` function for net.Error

Reply:
All code uses errors pkg now.

Reply:
great